### DISCUSSION TITLE
Pusbullet 1.2  notification problem when armed.

### DISCUSSION THREAD

**Discussion by l33d** [taligentx/dscKeybusInterface#85] [dscKeybusInterface Discussion]
Hi,

In new sketch i have this issue too. After armed system i received multiple notification for about 9 min like "Security system initializing"and "Security system connected". The problem isn`t  wifi, because when alarm is disarmed, and i disable and connect wifi, notification isn`t send. Next problem which i see now, notification about stay mode didn`t send, disarm notification too. 

_Originally posted by @l33d in https://github.com/taligentx/dscKeybusInterface/issues/55#issuecomment-498384888_

**Comment by taligentx** [comment]
Hi @l33d - I can't replicate this, I just re-tested and the initialization and connected messages are sent once at startup and the stay arm and disarm notifications are being sent.

If you're seeing multiple notifications about initializing, check the Serial output to see what's happening - the esp8266 could be resetting for some reason.  You can also run the `Status` example and watch the output to verify it matches the system states.

**Comment by l33d** [comment]
After this problem, when i testing pushbullet today and try use virtualblynk, virtualblynk stop responding for  button keys, like write pin is not connected( works for about 2 weeks ok when i try it sometimes). Maybe the problem is power?  I use stepdown converter on lm2557 from aliexpress. This problem is only mine?

Another question, can i restart esp with pusbullet or virtualblynk from app or remote? or must use relay and control it from another esp?

**Comment by taligentx** [comment]
Could be power, you're seeing issues that I can't replicate and I'm not seeing other similar issues being posted.

Pushbullet is one-way notification, but you could probably setup something in Blynk to restart the ESP though that's outside the scope of this project.

**Comment by l33d** [comment]
Thanks man. You`re probably right or maybe the problem is in my translate? I connect to serial and arm alarm, when alarm is armed esp always reset, but voltage is  ok, about 3,43v and when i turn it a bit high about 3,63v problem is not solved. I will check cables and more. When i find the issue i write about it.

When alarm is armed and esp reseting i see this output via console

> Initialization push notification sent successfully.
>  System DSC magistrali Keybus jest podłączony.
> 
> Exception (3):
> epc1=0x4020e128 epc2=0x00000000 epc3=0x00000000 excvaddr=0x40000000 depc=0x00000000
> 
> ctx: cont 
> sp: 3ffffd40 end: 3fffffd0 offset: 01a0
> 
> >>>stack>>>
> 3ffffee0:  7a797877 00000000 40106d34 00000c49  
> 3ffffef0:  3fffff6a 00000001 3fffff6b 3ffefde0  
> 3fffff00:  000e3549 00000000 00000001 00000000  
> 3fffff10:  00000001 3ffef589 3fffff6a 40207611  
> 3fffff20:  3fffdad0 00000000 3ffef550 40207628  
> 3fffff30:  3fffdad0 00000000 3ffef550 40202b93  
> 3fffff40:  74737953 44206d65 57204353 85c482c5  
> 3fffff50:  6e657a63 4e206569 414d524f 3a454e4c  
> 3fffff60:  646f7020 74737973 31316d65 31313131  
> 3fffff70:  31313131 31313131 31313131 31313131  
> 3fffff80:  31313131 31313131 31313131 31313131  
> 3fffff90:  31313131 31313131 31313131 31313131  
> 3fffffa0:  31313131 31313131 31313131 31313131  
> 3fffffb0:  31313131 31313131 31313131 31313131  
> 3fffffc0:  31313131 31313131 31313131 31313131  
> <<<stack<<<
> 
>  ets Jan  8 2013,rst cause:2, boot mode:(3,6)
> 
> load 0x4010f000, len 1384, room 16 
> tail 8
> chksum 0x2d
> csum 0x2d
> vbb28d4a3
> ~ld
> 
> 
> WiFi connected: 192.168.1.44
> Initialization push notification sent successfully.
>  System DSC magistrali Keybus jest podłączony.
> 

I try also connect ch_PD to vcc, but the problem didn`t disappear. I must try orginal sketch.


Edit:3 
Hahaha, probably i find the problem. I use my translate sketch with other esp on nodemcu with external power from powerbank and all is ok, when i try power nodemcu with alarm power via lm25xx but take from my one  zone expander the problem appear again.. But i don`t know why when i measure votlage all is ok. I change supply voltage and try again, but i think the problem is gone ;) This will be extreamly stupid issue :)

**Comment by l33d** [comment]
Ok, problem is probably in my sketch. When i armed stay i don`t received information about that. I upload original sketch and now i see this notification. Strange. I upload my translate sketch but i don`t know where is the problem. Power is other question, but now i connect it to alarm with original sketch and all is ok.

> void setup() {
>   Serial.begin(115200);
>   Serial.println();
>   Serial.println();
> 
>   WiFi.mode(WIFI_STA);
>   WiFi.begin(wifiSSID, wifiPassword);
>   while (WiFi.status() != WL_CONNECTED) delay(100);
>   Serial.print(F("WiFi connected: "));
>   Serial.println(WiFi.localIP());
> 
>   // Sends a push notification on startup to verify connectivity
>   #ifdef AXTLS_DEPRECATED
>     pushClient.setInsecure();  // Sets authentication method for BearSSL in esp8266 Arduino Core 2.5.0+
>   #endif
>   if (sendPush("System DSC uruchamianie")) Serial.println(F("Initialization push notification sent successfully."));
>   else Serial.println(F("Initialization push notification failed to send."));
> 
>   // Starts the Keybus interface
>   dsc.begin();
> 
>   Serial.println(F(" System DSC magistrali Keybus jest podłączony."));
> }
> 
> 
> void loop() {
>   if (dsc.handlePanel() && dsc.statusChanged) {  // Processes data only when a valid Keybus command has been read
>     dsc.statusChanged = false;                   // Resets the status flag
> 
>     // If the Keybus data buffer is exceeded, the sketch is too busy to process all Keybus commands.  Call
>     // handlePanel() more often, or increase dscBufferSize in the library: src/dscKeybusInterface.h
>     if (dsc.bufferOverflow) Serial.println(F("Przepełnienie bufora magistrali KEYBUS"));
>     dsc.bufferOverflow = false;
> 
>     // Checks if the interface is connected to the Keybus
>     if (dsc.keybusChanged) {
>       dsc.keybusChanged = false;  // Resets the Keybus data status flag
>       if (dsc.keybusConnected) sendPush("System DSC podłączony");
>       else sendPush("System DSC odłączony");
>     }
> 
>     // Checks status per partition
>     for (byte partition = 0; partition < dscPartitions; partition++) {
> 
>       // Checks armed status
>       if (dsc.armedChanged[partition]) {
>         dsc.armedChanged[partition] = false;  // Resets the partition armed status flag
>         if (dsc.armed[partition]) {
> 
>           char pushMessage[40] = "System DSC ";
>           if (dsc.armedAway[partition]) {
>             char armedState[31] = "Włączenie DOMOWE: podsystem ";
>             strcat(pushMessage, armedState);
>           }
>           else if (dsc.armedStay[partition]) {
>             char armedState[33] = "Włączenie NORMALNE: podsystem ";
>             strcat(pushMessage, armedState);
>           }
>           char partitionNumber[2];
>           itoa(partition + 1, partitionNumber, 10);
>           strcat(pushMessage, partitionNumber);
>           sendPush(pushMessage);
> 
>         }
>         else {
>           char pushMessage[34] = "System DSC rozbrojony: podsystem ";
>           char partitionNumber[2];
>           itoa(partition + 1, partitionNumber, 10);
>           strcat(pushMessage, partitionNumber);
>           sendPush(pushMessage);
>         }
>       }
> 
>       // Checks alarm triggered status
>       if (dsc.alarmChanged[partition]) {
>         dsc.alarmChanged[partition] = false;  // Resets the partition alarm status flag
> 
>         char pushMessage[34] = "System DSC w alarmie: podsystem ";
>         char partitionNumber[2];
>         itoa(partition + 1, partitionNumber, 10);
>         strcat(pushMessage, partitionNumber);
> 
>         if (dsc.alarm[partition]) sendPush(pushMessage);
>         else sendPush("System DSC rozbrojony po alarmie");
>       }
> 
>       // Checks fire alarm status
>       if (dsc.fireChanged[partition]) {
>         dsc.fireChanged[partition] = false;  // Resets the fire status flag
> 
>         char pushMessage[42] = "System DSC alarm POŻAROWY: podsystem ";
>         char partitionNumber[2];
>         itoa(partition + 1, partitionNumber, 10);
>         strcat(pushMessage, partitionNumber);
> 
>         if (dsc.fire[partition]) sendPush(pushMessage);
>         else sendPush("System DSC koniec alarmu POŻAROWEGO");
>       }
>     }
> 
>     // Checks for zones in alarm
>     // Zone alarm status is stored in the alarmZones[] and alarmZonesChanged[] arrays using 1 bit per zone, up to 64 zones
>     //   alarmZones[0] and alarmZonesChanged[0]: Bit 0 = Zone 1 ... Bit 7 = Zone 8
>     //   alarmZones[1] and alarmZonesChanged[1]: Bit 0 = Zone 9 ... Bit 7 = Zone 16
>     //   ...
>     //   alarmZones[7] and alarmZonesChanged[7]: Bit 0 = Zone 57 ... Bit 7 = Zone 64
>     if (dsc.alarmZonesStatusChanged) {
>       dsc.alarmZonesStatusChanged = false;                           // Resets the alarm zones status flag
>       for (byte zoneGroup = 0; zoneGroup < dscZones; zoneGroup++) {
>         for (byte zoneBit = 0; zoneBit < 8; zoneBit++) {
>           if (bitRead(dsc.alarmZonesChanged[zoneGroup], zoneBit)) {  // Checks an individual alarm zone status flag
>             bitWrite(dsc.alarmZonesChanged[zoneGroup], zoneBit, 0);  // Resets the individual alarm zone status flag
>             if (bitRead(dsc.alarmZones[zoneGroup], zoneBit)) {       // Zone alarm
>               char pushMessage[24] = "Alarm linia: ";
>               char zoneNumber[3];
>               itoa((zoneBit + 1 + (zoneGroup * 8)), zoneNumber, 10); // Determines the zone number
>               strcat(pushMessage, zoneNumber);
>               sendPush(pushMessage);
>             }
>             else {
>               char pushMessage[33] = "Powrót po alarmie linii: ";
>               char zoneNumber[3];
>               itoa((zoneBit + 1 + (zoneGroup * 8)), zoneNumber, 10); // Determines the zone number
>               strcat(pushMessage, zoneNumber);
>               sendPush(pushMessage);
>             }
>           }
>         }
>       }
>     }
> 
>     // Checks for AC power status
>     if (dsc.powerChanged) {
>       dsc.powerChanged = false;  // Resets the battery trouble status flag
>       if (dsc.powerTrouble) sendPush("Brak zasilania z sieci AC");
>       else sendPush("Zasilanie z sieci AC wróciło");
>     }
> 
>     // Checks for keypad fire alarm status
>     if (dsc.keypadFireAlarm) {
>       dsc.keypadFireAlarm = false;  // Resets the keypad fire alarm status flag
>       sendPush("DSC Klawiatura - wciśnięto POŻAR");
>     }
> 
>     // Checks for keypad aux auxiliary alarm status
>     if (dsc.keypadAuxAlarm) {
>       dsc.keypadAuxAlarm = false;  // Resets the keypad auxiliary alarm status flag
>       sendPush("DSC Klawiatura - wciśnięto POMOC");
>     }
> 
>     // Checks for keypad panic alarm status
>     if (dsc.keypadPanicAlarm) {
>       dsc.keypadPanicAlarm = false;  // Resets the keypad panic alarm status flag
>       sendPush("DSC Klawiatura - wciśnięto PANIKĘ");
>     }
>   }
> }
> 
> 
> bool sendPush(const char* pushMessage) {
> 
>   // Connects and sends the message as JSON
>   if (!pushClient.connect("api.pushbullet.com", 443)) return false;
>   pushClient.println(F("POST /v2/pushes HTTP/1.1"));
>   pushClient.println(F("Host: api.pushbullet.com"));
>   pushClient.println(F("User-Agent: ESP8266"));
>   pushClient.println(F("Accept: */*"));
>   pushClient.println(F("Content-Type: application/json"));
>   pushClient.print(F("Content-Length: "));
>   pushClient.println(strlen(pushMessage) + 25);  // Length including JSON data
>   pushClient.print(F("Access-Token: "));
>   pushClient.println(pushToken);
>   pushClient.println();
>   pushClient.print(F("{\"body\":\""));
>   pushClient.print(pushMessage);
>   pushClient.print(F("\",\"type\":\"note\"}"));
> 
>   // Waits for a response
>   unsigned long previousMillis = millis();
>   while (!pushClient.available()) {
>     dsc.handlePanel();
>     if (millis() - previousMillis > 3000) {
>       Serial.println(F("Connection timed out waiting for a response."));
>       pushClient.stop();
>       return false;
>     }
>     yield();
>   }
> 
>   // Reads the response until the first space - the next characters will be the HTTP status code
>   while (pushClient.available()) {
>     if (pushClient.read() == ' ') break;
>   }
> 
>   // Checks the first character of the HTTP status code - the message was sent successfully if the status code
>   // begins with "2"
>   char statusCode = pushClient.read();
> 
>   // Successful, reads the remaining response to clear the client buffer
>   if (statusCode == '2') {
>     while (pushClient.available()) pushClient.read();
>     pushClient.stop();
>     return true;
>   }
> 
>   // Unsuccessful, prints the response to serial to help debug
>   else {
>     Serial.println(F("Push notification error, response:"));
>     Serial.print(statusCode);
>     while (pushClient.available()) Serial.print((char)pushClient.read());
>     Serial.println();
>     pushClient.stop();
>     return false;
>   }
> }

**Comment by l33d** [comment]
Ok i test original sketch on nodemcu connected to my alarm for few days and it seems to be ok. After arm i received only one notification about partition and arm mode. What`s wrong is my translate sketch?

**Comment by taligentx** [comment]
Good to hear it's working with the original example sketch.  I recommend changing a few sections at a time to see which of your changes is causing the reset.

**Comment by l33d** [comment]
The problem is same as 1.0 sketch. I translate it too, and have this same problem with this same information as i remember. Do you know how can i translate the sketch properly? Can i use my own ascii charset like ą, ę, ł without problem? First i update arduino program, maybe there is a problem when compiling sketch with my charset.

**Comment by taligentx** [comment]
Unfortunately I'm not familiar with using different charsets, as far as I've seen the Arduino IDE should support UTF-8 except for the serial terminal which is ASCII only.

**Comment by taligentx** [comment]
Check your `char pushMessage[]` arrays, they seem to be too short - should be total number of characters + 1 for the terminating null character.

**Comment by l33d** [comment]
Probably you have right, i change the arrays char pushMessage[44],  char armedState[34] and char pushMessage[34] and this messages i received after arm/disarm system, and no errors. I upload this sketch to esp in my alarm and check, but problem is probably solved in this issue.

**Comment by taligentx** [comment]
Great, good to hear.
